gcc/rust/hir/rust-ast-lower-implitem.h

```diff
@@ -240,28 +240,32 @@ class ASTLowerTraitItem : public ASTLoweringBase
     return resolver.translated;
   }
 
-  void visit (AST::TraitItemFunc &func) override
+  void visit (AST::Function &func) override
   {
-    AST::TraitFunctionDecl &ref = func.get_trait_function_decl ();
-
     std::vector<std::unique_ptr<HIR::WhereClauseItem> > where_clause_items;
     HIR::WhereClause where_clause (std::move (where_clause_items));
     HIR::FunctionQualifiers qualifiers
-      = lower_qualifiers (func.get_trait_function_decl ().get_qualifiers ());
+      = lower_qualifiers (func.get_qualifiers ());
 
     std::vector<std::unique_ptr<HIR::GenericParam> > generic_params;
-    if (ref.has_generics ())
+    if (func.has_generics ())
       {
-	generic_params = lower_generic_params (ref.get_generic_params ());
+	generic_params = lower_generic_params (func.get_generic_params ());
       }
 
     std::unique_ptr<HIR::Type> return_type
-      = ref.has_return_type () ? std::unique_ptr<HIR::Type> (
-	  ASTLoweringType::translate (ref.get_return_type ().get ()))
-			       : nullptr;
+      = func.has_return_type () ? std::unique_ptr<HIR::Type> (
+	  ASTLoweringType::translate (func.get_return_type ().get ()))
+				: nullptr;
+
+    // set self parameter to error if this is a method
+    // else lower to hir
+    HIR::SelfParam self_param = func.has_self_param ()
+				  ? lower_self (func.get_self_param ())
+				  : HIR::SelfParam::error ();
 
     std::vector<HIR::FunctionParam> function_params;
-    for (auto &p : ref.get_function_params ())
+    for (auto &p : func.get_function_params ())
```

Comment:
Maybe we should introduce a member function in AST to retrieve regular (non self|variadic) params ? I'm almost sure the current API might lead to some errors in the future. Would it be really useful ?

Should not be done in this PR anyway.